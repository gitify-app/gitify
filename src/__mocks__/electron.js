window.Notification = function (title) {
  this.title = title;

  return {
    onclick: jest.fn(),
  };
};

window.Audio = class Audio {
  constructor(path) {
    this.path = path;
  }

  play() {}
};

window.localStorage = {
  store: {},
  getItem: function (key) {
    return this.store[key];
  },
  setItem: function (key, item) {
    this.store[key] = item;
  },
  removeItem: jest.fn(),
};

window.alert = jest.fn();

module.exports = {
  ipcRenderer: {
    send: jest.fn(),
    on: jest.fn(),
    sendSync: jest.fn(),
    invoke: jest.fn((channel, ...args) => {
      switch (channel) {
        case 'get-platform':
          return Promise.resolve('darwin');
        case 'gitify:version':
          return Promise.resolve('0.0.1');
        default:
          return Promise.reject(new Error(`Unknown channel: ${channel}`));
      }
    }),
  },
  shell: {
    openExternal: jest.fn(),
  },
};
