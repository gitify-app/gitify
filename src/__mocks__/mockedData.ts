import type {
  AccountNotifications,
  EnterpriseAccount,
  GitifyUser,
} from '../types';
import type {
  Discussion,
  DiscussionComments,
  GraphQLSearch,
  Notification,
  Repository,
  User,
} from '../typesGithub';
import Constants from '../utils/constants';

export const mockedEnterpriseAccounts: EnterpriseAccount[] = [
  {
    hostname: 'github.gitify.io',
    token: '1234568790',
  },
];

export const mockedUser: GitifyUser = {
  login: 'octocat',
  name: 'Mona Lisa Octocat',
  id: 123456789,
};

export const mockedNotificationUser: User = {
  login: 'octocat',
  id: 123456789,
  node_id: 'MDQ6VXNlcjE=',
  avatar_url: 'https://avatars.githubusercontent.com/u/583231?v=4',
  gravatar_url: '',
  url: 'https://api.github.com/users/octocat',
  html_url: 'https://github.com/octocat',
  followers_url: 'https://api.github.com/users/octocat/followers',
  following_url: 'https://api.github.com/users/octocat/following{/other_user}',
  gists_url: 'https://api.github.com/users/octocat/gists{/gist_id}',
  starred_url: 'https://api.github.com/users/octocat/starred{/owner}{/repo}',
  subscriptions_url: 'https://api.github.com/users/octocat/subscriptions',
  organizations_url: 'https://api.github.com/users/octocat/orgs',
  repos_url: 'https://api.github.com/users/octocat/repos',
  events_url: 'https://api.github.com/users/octocat/events{/privacy}',
  received_events_url: 'https://api.github.com/users/octocat/received_events',
  type: 'User',
  site_admin: false,
};

export const mockedCommenterUser: User = {
  login: 'dummyuser',
  id: 987654321,
  node_id: 'MDQ6VXNlcjI=',
  avatar_url: 'https://avatars.githubusercontent.com/u/123456789?v=4',
  gravatar_url: '',
  url: 'https://api.github.com/users/dummyuser',
  html_url: 'https://github.com/dummyuser',
  followers_url: 'https://api.github.com/users/dummyuser/followers',
  following_url:
    'https://api.github.com/users/dummyuser/following{/other_user}',
  gists_url: 'https://api.github.com/users/dummyuser/gists{/gist_id}',
  starred_url: 'https://api.github.com/users/dummyuser/starred{/owner}{/repo}',
  subscriptions_url: 'https://api.github.com/users/dummyuser/subscriptions',
  organizations_url: 'https://api.github.com/users/dummyuser/orgs',
  repos_url: 'https://api.github.com/users/dummyuser/repos',
  events_url: 'https://api.github.com/users/dummyuser/events{/privacy}',
  received_events_url: 'https://api.github.com/users/dummyuser/received_events',
  type: 'User',
  site_admin: false,
};

export const mockedSingleNotification: Notification = {
  hostname: Constants.DEFAULT_AUTH_OPTIONS.hostname,
  id: '138661096',
  unread: true,
  reason: 'subscribed',
  updated_at: '2017-05-20T17:51:57Z',
  last_read_at: '2017-05-20T17:06:51Z',
  subject: {
    title: 'I am a robot and this is a test!',
    url: 'https://api.github.com/repos/manosim/notifications-test/issues/1',
    latest_comment_url:
      'https://api.github.com/repos/manosim/notifications-test/issues/comments/302888448',
    type: 'Issue',
    state: 'open',
    user: {
      login: 'manosim',
      html_url: 'https://github.com/manosim',
      avatar_url: 'https://avatars0.githubusercontent.com/u/6333409?v=3',
      type: 'User',
    },
  },
  repository: {
    id: 57216596,
    node_id: 'MDEwOlJlcG9zaXRvcnkzNjAyOTcwNg==',
    name: 'notifications-test',
    full_name: 'manosim/notifications-test',
    url: 'https://api.github.com/manosim/notifications-test',
    owner: {
      login: 'manosim',
      id: 6333409,
      node_id: 'MDQ6VXNlcjYzMzM0MDk=',
      avatar_url: 'https://avatars0.githubusercontent.com/u/6333409?v=3',
      gravatar_id: '',
      url: 'https://api.github.com/users/manosim',
      html_url: 'https://github.com/manosim',
      followers_url: 'https://api.github.com/users/manosim/followers',
      following_url:
        'https://api.github.com/users/manosim/following{/other_user}',
      gists_url: 'https://api.github.com/users/manosim/gists{/gist_id}',
      starred_url:
        'https://api.github.com/users/manosim/starred{/owner}{/repo}',
      subscriptions_url: 'https://api.github.com/users/manosim/subscriptions',
      organizations_url: 'https://api.github.com/users/manosim/orgs',
      repos_url: 'https://api.github.com/users/manosim/repos',
      events_url: 'https://api.github.com/users/manosim/events{/privacy}',
      received_events_url:
        'https://api.github.com/users/manosim/received_events',
      type: 'User',
      site_admin: false,
    },
    private: true,
    description: 'Test Repository',
    fork: false,
    archive_url:
      'https://api.github.com/repos/manosim/notifications-test/{archive_format}{/ref}',
    assignees_url:
      'https://api.github.com/repos/manosim/notifications-test/assignees{/user}',
    blobs_url:
      'https://api.github.com/repos/manosim/notifications-test/git/blobs{/sha}',
    branches_url:
      'https://api.github.com/repos/manosim/notifications-test/branches{/branch}',
    collaborators_url:
      'https://api.github.com/repos/manosim/notifications-test/collaborators{/collaborator}',
    comments_url:
      'https://api.github.com/repos/manosim/notifications-test/comments{/number}',
    commits_url:
      'https://api.github.com/repos/manosim/notifications-test/commits{/sha}',
    compare_url:
      'https://api.github.com/repos/manosim/notifications-test/compare/{base}...{head}',
    contents_url:
      'https://api.github.com/repos/manosim/notifications-test/contents/{+path}',
    contributors_url:
      'https://api.github.com/repos/manosim/notifications-test/contributors',
    deployments_url:
      'https://api.github.com/repos/manosim/notifications-test/deployments',
    downloads_url:
      'https://api.github.com/repos/manosim/notifications-test/downloads',
    events_url:
      'https://api.github.com/repos/manosim/notifications-test/events',
    forks_url: 'https://api.github.com/repos/manosim/notifications-test/forks',
    git_commits_url:
      'https://api.github.com/repos/manosim/notifications-test/git/commits{/sha}',
    git_refs_url:
      'https://api.github.com/repos/manosim/notifications-test/git/refs{/sha}',
    git_tags_url:
      'https://api.github.com/repos/manosim/notifications-test/git/tags{/sha}',
    hooks_url: 'https://api.github.com/repos/manosim/notifications-test/hooks',
    html_url: 'https://github.com/manosim/notifications-test',
    issue_comment_url:
      'https://api.github.com/repos/manosim/notifications-test/issues/comments{/number}',
    issue_events_url:
      'https://api.github.com/repos/manosim/notifications-test/issues/events{/number}',
    issues_url:
      'https://api.github.com/repos/manosim/notifications-test/issues{/number}',
    keys_url:
      'https://api.github.com/repos/manosim/notifications-test/keys{/key_id}',
    labels_url:
      'https://api.github.com/repos/manosim/notifications-test/labels{/name}',
    languages_url:
      'https://api.github.com/repos/manosim/notifications-test/languages',
    merges_url:
      'https://api.github.com/repos/manosim/notifications-test/merges',
    milestones_url:
      'https://api.github.com/repos/manosim/notifications-test/milestones{/number}',
    notifications_url:
      'https://api.github.com/repos/manosim/notifications-test/notifications{?since,all,participating}',
    pulls_url:
      'https://api.github.com/repos/manosim/notifications-test/pulls{/number}',
    releases_url:
      'https://api.github.com/repos/manosim/notifications-test/releases{/id}',
    stargazers_url:
      'https://api.github.com/repos/manosim/notifications-test/stargazers',
    statuses_url:
      'https://api.github.com/repos/manosim/notifications-test/statuses/{sha}',
    subscribers_url:
      'https://api.github.com/repos/manosim/notifications-test/subscribers',
    subscription_url:
      'https://api.github.com/repos/manosim/notifications-test/subscription',
    tags_url: 'https://api.github.com/repos/manosim/notifications-test/tags',
    teams_url: 'https://api.github.com/repos/manosim/notifications-test/teams',
    trees_url:
      'https://api.github.com/repos/manosim/notifications-test/git/trees{/sha}',
  },
  url: 'https://api.github.com/notifications/threads/138661096',
  subscription_url:
    'https://api.github.com/notifications/threads/138661096/subscription',
};

// 2 Notifications
// Repository : 'manosim/notifications-tests'
export const mockedGithubNotifications = [
  mockedSingleNotification,
  {
    id: '148827438',
    unread: true,
    reason: 'author',
    updated_at: '2017-05-20T17:06:34Z',
    last_read_at: '2017-05-20T16:59:03Z',
    subject: {
      title: 'Improve the UI',
      url: 'https://api.github.com/repos/manosim/notifications-test/issues/4',
      latest_comment_url:
        'https://api.github.com/repos/manosim/notifications-test/issues/comments/302885965',
      type: 'Issue',
    },
    repository: {
      id: 57216596,
      name: 'notifications-test',
      full_name: 'manosim/notifications-test',
      owner: {
        login: 'manosim',
        id: 6333409,
        avatar_url: 'https://avatars0.githubusercontent.com/u/6333409?v=3',
        gravatar_id: '',
        url: 'https://api.github.com/users/manosim',
        html_url: 'https://github.com/manosim',
        followers_url: 'https://api.github.com/users/manosim/followers',
        following_url:
          'https://api.github.com/users/manosim/following{/other_user}',
        gists_url: 'https://api.github.com/users/manosim/gists{/gist_id}',
        starred_url:
          'https://api.github.com/users/manosim/starred{/owner}{/repo}',
        subscriptions_url: 'https://api.github.com/users/manosim/subscriptions',
        organizations_url: 'https://api.github.com/users/manosim/orgs',
        repos_url: 'https://api.github.com/users/manosim/repos',
        events_url: 'https://api.github.com/users/manosim/events{/privacy}',
        received_events_url:
          'https://api.github.com/users/manosim/received_events',
        type: 'User',
        site_admin: false,
      },
      private: true,
      html_url: 'https://github.com/manosim/notifications-test',
      description: null,
      fork: false,
      // Removed the rest of the properties
    } as unknown as Repository,
    url: 'https://api.github.com/notifications/threads/148827438',
    subscription_url:
      'https://api.github.com/notifications/threads/148827438/subscription',
    hostname: Constants.DEFAULT_AUTH_OPTIONS.hostname,
  } as Notification,
];

// 2 Notifications
// Repository : 'myorg/notifications-test'
export const mockedEnterpriseNotifications = [
  {
    id: '4',
    unread: true,
    reason: 'subscribed',
    updated_at: '2017-05-20T13:02:48Z',
    last_read_at: null,
    subject: {
      title: 'Release 0.0.1',
      url: 'https://github.gitify.io/api/v3/repos/myorg/notifications-test/releases/1',
      latest_comment_url:
        'https://github.gitify.io/api/v3/repos/myorg/notifications-test/releases/1',
      type: 'Release',
    },
    repository: {
      id: 1,
      name: 'notifications-test',
      full_name: 'myorg/notifications-test',
      owner: {
        login: 'myorg',
        id: 4,
        avatar_url: 'https://github.gitify.io/avatars/u/4?',
        gravatar_id: '',
        url: 'https://github.gitify.io/api/v3/users/myorg',
        html_url: 'https://github.gitify.io/myorg',
        followers_url: 'https://github.gitify.io/api/v3/users/myorg/followers',
        following_url:
          'https://github.gitify.io/api/v3/users/myorg/following{/other_user}',
        gists_url:
          'https://github.gitify.io/api/v3/users/myorg/gists{/gist_id}',
        starred_url:
          'https://github.gitify.io/api/v3/users/myorg/starred{/owner}{/repo}',
        subscriptions_url:
          'https://github.gitify.io/api/v3/users/myorg/subscriptions',
        organizations_url: 'https://github.gitify.io/api/v3/users/myorg/orgs',
        repos_url: 'https://github.gitify.io/api/v3/users/myorg/repos',
        events_url:
          'https://github.gitify.io/api/v3/users/myorg/events{/privacy}',
        received_events_url:
          'https://github.gitify.io/api/v3/users/myorg/received_events',
        type: 'Organization',
        site_admin: false,
      },
      private: true,
      html_url: 'https://github.gitify.io/myorg/notifications-test',
      description: null,
      fork: false,
      // Removed the rest of the properties
    } as unknown as Repository,
    url: 'https://github.gitify.io/api/v3/notifications/threads/4',
    subscription_url:
      'https://github.gitify.io/api/v3/notifications/threads/4/subscription',
  } as Notification,
  {
    id: '3',
    unread: true,
    reason: 'subscribed',
    updated_at: '2017-05-20T15:52:20Z',
    last_read_at: '2017-05-20T14:20:55Z',
    subject: {
      title: 'Bump Version',
      url: 'https://github.gitify.io/api/v3/repos/myorg/notifications-test/pulls/3',
      latest_comment_url:
        'https://github.gitify.io/api/v3/repos/myorg/notifications-test/issues/comments/21',
      type: 'PullRequest',
    },
    repository: {
      id: 1,
      name: 'notifications-test',
      full_name: 'myorg/notifications-test',
      owner: {
        login: 'myorg',
        id: 4,
        avatar_url: 'https://github.gitify.io/avatars/u/4?',
        gravatar_id: '',
        url: 'https://github.gitify.io/api/v3/users/myorg',
        html_url: 'https://github.gitify.io/myorg',
        followers_url: 'https://github.gitify.io/api/v3/users/myorg/followers',
        following_url:
          'https://github.gitify.io/api/v3/users/myorg/following{/other_user}',
        gists_url:
          'https://github.gitify.io/api/v3/users/myorg/gists{/gist_id}',
        starred_url:
          'https://github.gitify.io/api/v3/users/myorg/starred{/owner}{/repo}',
        subscriptions_url:
          'https://github.gitify.io/api/v3/users/myorg/subscriptions',
        organizations_url: 'https://github.gitify.io/api/v3/users/myorg/orgs',
        repos_url: 'https://github.gitify.io/api/v3/users/myorg/repos',
        events_url:
          'https://github.gitify.io/api/v3/users/myorg/events{/privacy}',
        received_events_url:
          'https://github.gitify.io/api/v3/users/myorg/received_events',
        type: 'Organization',
        site_admin: false,
      },
      private: true,
      html_url: 'https://github.gitify.io/myorg/notifications-test',
      description: null,
      fork: false,
      // Removed the rest of the properties
    } as unknown as Repository,
    url: 'https://github.gitify.io/api/v3/notifications/threads/3',
    subscription_url:
      'https://github.gitify.io/api/v3/notifications/threads/3/subscription',
    hostname: Constants.DEFAULT_AUTH_OPTIONS.hostname,
  } as Notification,
];

export const mockedAccountNotifications: AccountNotifications[] = [
  {
    hostname: 'github.com',
    notifications: mockedGithubNotifications,
  },
  {
    hostname: 'github.gitify.io',
    notifications: mockedEnterpriseNotifications,
  },
];

export const mockedSingleAccountNotifications: AccountNotifications[] = [
  {
    hostname: 'github.com',
    notifications: [mockedSingleNotification],
  },
];

const mockDiscussionComments: DiscussionComments = {
  nodes: [
    {
      databaseId: 2215656,
      createdAt: '2022-02-20T18:33:39Z',
      author: {
        login: 'comment-user',
        url: 'https://github.com/comment-user',
        avatar_url: 'https://avatars.githubusercontent.com/u/123456789?v=4',
        type: 'User',
      },
      replies: {
        nodes: [],
      },
    },
    {
      databaseId: 2217789,
      createdAt: '2022-02-21T03:30:42Z',
      author: {
        login: 'comment-user',
        url: 'https://github.com/comment-user',
        avatar_url: 'https://avatars.githubusercontent.com/u/123456789?v=4',
        type: 'User',
      },
      replies: {
        nodes: [],
      },
    },
    {
      databaseId: 2223243,
      createdAt: '2022-02-21T18:26:27Z',
      author: {
        login: 'comment-user',
        url: 'https://github.com/comment-user',
        avatar_url: 'https://avatars.githubusercontent.com/u/123456789?v=4',
        type: 'User',
      },
      replies: {
        nodes: [
          {
            databaseId: 2232922,
            createdAt: '2022-02-23T00:57:58Z',
            author: {
              login: 'reply-user',
              url: 'https://github.com/reply-user',
              avatar_url:
                'https://avatars.githubusercontent.com/u/123456789?v=4',
              type: 'User',
            },
          },
        ],
      },
    },
    {
      databaseId: 2232921,
      createdAt: '2022-02-23T00:57:49Z',
      author: {
        login: 'comment-user',
        url: 'https://github.com/comment-user',
        avatar_url: 'https://avatars.githubusercontent.com/u/123456789?v=4',
        type: 'User',
      },
      replies: {
        nodes: [],
      },
    },
    {
      databaseId: 2258799,
      createdAt: '2022-02-27T01:22:20Z',
      author: {
        login: 'comment-user',
        url: 'https://github.com/comment-user',
        avatar_url: 'https://avatars.githubusercontent.com/u/123456789?v=4',
        type: 'User',
      },
      replies: {
        nodes: [
          {
            databaseId: 2300902,
            createdAt: '2022-03-05T17:43:52Z',
            author: {
              login: 'reply-user',
              url: 'https://github.com/reply-user',
              avatar_url:
                'https://avatars.githubusercontent.com/u/123456789?v=4',
              type: 'User',
            },
          },
        ],
      },
    },
    {
      databaseId: 2297637,
      createdAt: '2022-03-04T20:39:44Z',
      author: {
        login: 'comment-user',
        url: 'https://github.com/comment-user',
        avatar_url: 'https://avatars.githubusercontent.com/u/123456789?v=4',
        type: 'User',
      },

      replies: {
        nodes: [
          {
            databaseId: 2300893,
            createdAt: '2022-03-05T17:41:04Z',
            author: {
              login: 'reply-user',
              url: 'https://github.com/reply-user',
              avatar_url:
                'https://avatars.githubusercontent.com/u/123456789?v=4',
              type: 'User',
            },
          },
        ],
      },
    },
    {
      databaseId: 2299763,
      createdAt: '2022-03-05T11:05:42Z',
      author: {
        login: 'comment-user',
        url: 'https://github.com/comment-user',
        avatar_url: 'https://avatars.githubusercontent.com/u/123456789?v=4',
        type: 'User',
      },
      replies: {
        nodes: [
          {
            databaseId: 2300895,
            createdAt: '2022-03-05T17:41:44Z',
            author: {
              login: 'reply-user',
              url: 'https://github.com/reply-user',
              avatar_url:
                'https://avatars.githubusercontent.com/u/123456789?v=4',
              type: 'User',
            },
          },
        ],
      },
    },
  ],
};

export const mockedGraphQLResponse: GraphQLSearch<Discussion> = {
  data: {
    data: {
      search: {
        nodes: [
          {
            viewerSubscription: 'SUBSCRIBED',
            title: '1.16.0',
            isAnswered: false,
            stateReason: 'OPEN',
            url: 'https://github.com/manosim/notifications-test/discussions/612',
            author: {
              login: 'discussion-creator',
              url: 'https://github.com/discussion-creator',
              avatar_url:
                'https://avatars.githubusercontent.com/u/123456789?v=4',
              type: 'User',
            },
            comments: mockDiscussionComments,
          },
          {
            viewerSubscription: 'IGNORED',
            title: '1.16.0',
            isAnswered: false,
            stateReason: 'ANSWERED',
            url: 'https://github.com/manosim/notifications-test/discussions/123',
            author: {
              login: 'discussion-creator',
              url: 'https://github.com/discussion-creator',
              avatar_url:
                'https://avatars.githubusercontent.com/u/123456789?v=4',
              type: 'User',
            },
            comments: mockDiscussionComments,
          },
        ],
      },
    },
  },
};
