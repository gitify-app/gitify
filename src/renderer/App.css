html,
body,
#root {
  height: 100%;
  -webkit-user-select: none;
}

html::-webkit-scrollbar,
div.flex-grow::-webkit-scrollbar {
  width: 10px;
}

html::-webkit-scrollbar-thumb,
div.flex-grow::-webkit-scrollbar-thumb {
  background-color: #c1c1c1;
  border-radius: 10px;
}

html::-webkit-scrollbar-thumb:hover,
div.flex-grow::-webkit-scrollbar-thumb:hover {
  background-color: #a8a8a8;
  border-radius: 10px;
}

html.dark::-webkit-scrollbar-track,
html.dark div.flex-grow::-webkit-scrollbar-track {
  background-color: #090e15;
}

html.dark::-webkit-scrollbar-thumb,
html.dark div.flex-grow::-webkit-scrollbar-thumb {
  background-color: #24292e;
}

html.dark::-webkit-scrollbar-thumb:hover,
html.dark div.flex-grow::-webkit-scrollbar-thumb:hover {
  background-color: #3a3f44;
}

body {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
    "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
    sans-serif;
  margin: 0;
  cursor: default;
}

#nprogress {
  pointer-events: none;
}

#nprogress .bar {
  position: fixed;
  top: 0;
  left: 50px;
  z-index: 1031;
  background: #203354;

  width: 100%;
  height: 2px;
}

/**
 * Set emoji size according to surrounding text.
 * Ref: https://github.com/jdecked/twemoji?tab=readme-ov-file#inline-styles
 */
img.emoji {
  height: 1em;
  width: 1em;
  margin: 0 0.05em 0 0.1em;
  vertical-align: -0.1em;
}
