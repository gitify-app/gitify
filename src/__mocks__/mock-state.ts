import { type AuthState, type SettingsState, Theme } from '../types';
import { mockedEnterpriseAccounts, mockedUser } from './mockedData';

export const mockAccounts: AuthState = {
  token: 'token-123-456',
  enterpriseAccounts: mockedEnterpriseAccounts,
  user: mockedUser,
};

export const mockSettings: SettingsState = {
  participating: false,
  playSound: true,
  showNotifications: true,
  showBots: true,
  showNotificationsCountInTray: false,
  openAtStartup: false,
  theme: Theme.SYSTEM,
  detailedNotifications: false,
  markAsDoneOnOpen: false,
  showAccountHostname: false,
};
