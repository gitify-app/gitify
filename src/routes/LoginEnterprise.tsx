const ipcRenderer = require('electron').ipcRenderer;

import { ArrowLeftIcon } from '@primer/octicons-react';

import { type FC, useCallback, useContext, useEffect } from 'react';
import { Form, type FormRenderProps } from 'react-final-form';
import { useNavigate } from 'react-router-dom';

import { FieldInput } from '../components/fields/FieldInput';
import { AppContext } from '../context/App';
import type { AuthOptions } from '../types';

interface IValues {
  hostname?: string;
  clientId?: string;
  clientSecret?: string;
}

interface IFormErrors {
  hostname?: string;
  clientId?: string;
  clientSecret?: string;
}

export const validate = (values: IValues): IFormErrors => {
  const errors: IFormErrors = {};
  if (!values.hostname) {
    errors.hostname = 'Required';
  } else if (
    !/^([a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,6}$/i.test(
      values.hostname,
    )
  ) {
    errors.hostname = 'Invalid hostname.';
  }

  if (!values.clientId) {
    // 20
    errors.clientId = 'Required';
  } else if (!/^[A-Z0-9]{20}$/i.test(values.clientId)) {
    errors.clientId = 'Invalid client id.';
  }

  if (!values.clientSecret) {
    // 40
    errors.clientSecret = 'Required';
  } else if (!/^[A-Z0-9]{40}$/i.test(values.clientSecret)) {
    errors.clientSecret = 'Invalid client secret.';
  }

  return errors;
};

export const LoginEnterpriseRoute: FC = () => {
  const {
    accounts: { enterpriseAccounts },
    loginEnterprise,
  } = useContext(AppContext);
  const navigate = useNavigate();

  useEffect(() => {
    if (enterpriseAccounts.length) {
      ipcRenderer.send('reopen-window');
      navigate(-1);
    }
  }, [enterpriseAccounts]);

  const renderForm = (formProps: FormRenderProps) => {
    const { handleSubmit, submitting, pristine } = formProps;

    return (
      <form onSubmit={handleSubmit}>
        <FieldInput
          name="hostname"
          label="Hostname"
          placeholder="github.company.com"
        />

        <FieldInput name="clientId" label="Client ID" placeholder="123456789" />

        <FieldInput
          name="clientSecret"
          label="Client Secret"
          placeholder="ABC123DEF456"
        />

        <button
          className="float-right px-4 py-2 my-4 bg-gray-300 font-semibold rounded text-sm text-center hover:bg-gray-500 hover:text-white dark:text-black focus:outline-none"
          title="Login Button"
          disabled={submitting || pristine}
          type="submit"
        >
          Login
        </button>
      </form>
    );
  };

  const login = useCallback(async (data: IValues) => {
    try {
      await loginEnterprise(data as AuthOptions);
    } catch (err) {
      // Skip
    }
  }, []);

  return (
    <div className="flex-1 bg-white dark:bg-gray-dark dark:text-white">
      <div className="flex justify-between items-center mt-4 py-2 mx-8">
        <button
          type="button"
          className="focus:outline-none"
          title="Go Back"
          onClick={() => navigate(-1)}
        >
          <ArrowLeftIcon
            size={20}
            className="hover:text-gray-400"
            aria-label="Go Back"
          />
        </button>

        <h3 className="text-lg font-semibold">Login with GitHub Enterprise</h3>
      </div>

      <div className="flex-1 px-8">
        <Form
          initialValues={{
            hostname: '',
            clientId: '',
            clientSecret: '',
          }}
          onSubmit={login}
          validate={validate}
        >
          {renderForm}
        </Form>
      </div>
    </div>
  );
};
