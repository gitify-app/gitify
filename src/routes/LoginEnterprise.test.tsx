import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import * as TestRenderer from 'react-test-renderer';
const { ipcRenderer } = require('electron');
import { mockedEnterpriseAccounts } from '../__mocks__/mockedData';
import { AppContext } from '../context/App';
import type { AuthState } from '../types';
import { LoginEnterpriseRoute, validate } from './LoginEnterprise';

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('routes/LoginEnterprise.tsx', () => {
  const mockAccounts: AuthState = {
    enterpriseAccounts: [],
    user: null,
  };

  beforeEach(() => {
    mockNavigate.mockReset();

    jest.spyOn(ipcRenderer, 'send');
  });

  it('renders correctly', () => {
    const tree = TestRenderer.create(
      <AppContext.Provider value={{ accounts: mockAccounts }}>
        <MemoryRouter>
          <LoginEnterpriseRoute />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('let us go back', () => {
    render(
      <AppContext.Provider value={{ accounts: mockAccounts }}>
        <MemoryRouter>
          <LoginEnterpriseRoute />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    fireEvent.click(screen.getByLabelText('Go Back'));
    expect(mockNavigate).toHaveBeenNthCalledWith(1, -1);
  });

  it('should validate the form values', () => {
    const emptyValues = {
      hostname: null,
      clientId: null,
      clientSecret: null,
    };

    let values = {
      ...emptyValues,
    };
    expect(validate(values).hostname).toBe('Required');
    expect(validate(values).clientId).toBe('Required');
    expect(validate(values).clientSecret).toBe('Required');

    values = {
      ...emptyValues,
      hostname: 'hello',
      clientId: '!@£INVALID-.1',
      clientSecret: '!@£INVALID-.1',
    };
    expect(validate(values).hostname).toBe('Invalid hostname.');
    expect(validate(values).clientId).toBe('Invalid client id.');
    expect(validate(values).clientSecret).toBe('Invalid client secret.');
  });

  it('should receive a logged-in enterprise account', () => {
    const { rerender } = render(
      <AppContext.Provider value={{ accounts: mockAccounts }}>
        <MemoryRouter>
          <LoginEnterpriseRoute />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    rerender(
      <AppContext.Provider
        value={{
          accounts: {
            enterpriseAccounts: mockedEnterpriseAccounts,
            user: null,
          },
        }}
      >
        <MemoryRouter>
          <LoginEnterpriseRoute />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    expect(ipcRenderer.send).toHaveBeenCalledTimes(1);
    expect(ipcRenderer.send).toHaveBeenCalledWith('reopen-window');
    expect(mockNavigate).toHaveBeenNthCalledWith(1, -1);
  });

  it('should render the form with errors', () => {
    render(
      <AppContext.Provider value={{ accounts: mockAccounts }}>
        <MemoryRouter>
          <LoginEnterpriseRoute />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    fireEvent.change(screen.getByLabelText('Hostname'), {
      target: { value: 'test' },
    });
    fireEvent.change(screen.getByLabelText('Client ID'), {
      target: { value: '123' },
    });
    fireEvent.change(screen.getByLabelText('Client Secret'), {
      target: { value: 'abc' },
    });

    fireEvent.submit(screen.getByTitle('Login Button'));

    expect(screen.getByText('Invalid hostname.')).toBeTruthy();
    expect(screen.getByText('Invalid client id.')).toBeTruthy();
    expect(screen.getByText('Invalid client secret.')).toBeTruthy();
  });
});
