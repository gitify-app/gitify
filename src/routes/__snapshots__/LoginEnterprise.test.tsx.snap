// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes/LoginEnterprise.tsx renders correctly 1`] = `
<div
  className="flex-1 bg-white dark:bg-gray-dark dark:text-white"
>
  <div
    className="flex justify-between items-center mt-4 py-2 mx-8"
  >
    <button
      className="focus:outline-none"
      onClick={[Function]}
      title="Go Back"
      type="button"
    >
      <svg
        aria-label="Go Back"
        className="hover:text-gray-400"
        fill="currentColor"
        focusable="false"
        height={20}
        role="img"
        style={
          {
            "display": "inline-block",
            "overflow": "visible",
            "userSelect": "none",
            "verticalAlign": "text-bottom",
          }
        }
        viewBox="0 0 16 16"
        width={20}
      >
        <path
          d="M7.78 12.53a.75.75 0 0 1-1.06 0L2.47 8.28a.75.75 0 0 1 0-1.06l4.25-4.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042L4.81 7h7.44a.75.75 0 0 1 0 1.5H4.81l2.97 2.97a.75.75 0 0 1 0 1.06Z"
        />
      </svg>
    </button>
    <h3
      className="text-lg font-semibold"
    >
      Login with GitHub Enterprise
    </h3>
  </div>
  <div
    className="flex-1 px-8"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="mb-4"
      >
        <label
          className="block tracking-wide text-grey-dark text-sm font-semibold mb-2"
          htmlFor="hostname"
        >
          Hostname
        </label>
        <input
          className="appearance-none block w-full dark:text-gray-800 bg-gray-100 border border-red rounded py-1.5 px-4 mb-2 focus:bg-gray-200 focus:outline-none text-sm"
          id="hostname"
          name="hostname"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="github.company.com"
          required={false}
          type="text"
          value=""
        />
      </div>
      <div
        className="mb-4"
      >
        <label
          className="block tracking-wide text-grey-dark text-sm font-semibold mb-2"
          htmlFor="clientId"
        >
          Client ID
        </label>
        <input
          className="appearance-none block w-full dark:text-gray-800 bg-gray-100 border border-red rounded py-1.5 px-4 mb-2 focus:bg-gray-200 focus:outline-none text-sm"
          id="clientId"
          name="clientId"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="123456789"
          required={false}
          type="text"
          value=""
        />
      </div>
      <div
        className="mb-4"
      >
        <label
          className="block tracking-wide text-grey-dark text-sm font-semibold mb-2"
          htmlFor="clientSecret"
        >
          Client Secret
        </label>
        <input
          className="appearance-none block w-full dark:text-gray-800 bg-gray-100 border border-red rounded py-1.5 px-4 mb-2 focus:bg-gray-200 focus:outline-none text-sm"
          id="clientSecret"
          name="clientSecret"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="ABC123DEF456"
          required={false}
          type="text"
          value=""
        />
      </div>
      <button
        className="float-right px-4 py-2 my-4 bg-gray-300 font-semibold rounded text-sm text-center hover:bg-gray-500 hover:text-white dark:text-black focus:outline-none"
        disabled={true}
        title="Login Button"
        type="submit"
      >
        Login
      </button>
    </form>
  </div>
</div>
`;
